lib/puppet/provider/network/libvirt.rb

```diff
@@ -0,0 +1,177 @@
+require 'libvirt'
+require 'erb'
+require 'nokogiri'
+
+Puppet::Type.type(:network).provide(:libvirt) do
+  desc "Create domains with libvirt"
+  
+  mk_resource_methods
+
+  $conn = Libvirt::open('qemu:///system')
+
+  def self.parse_network(network)
+    doc = Nokogiri::XML(network.xml_desc)
+    definition = {}
+    definition[:name] = doc.at_xpath('//name').content
+    definition[:uuid] = doc.at_xpath('//uuid').content
+    if doc.at_xpath('//bridge') and doc.at_xpath('//bridge').attribute('name')
+      definition[:bridge] = doc.at_xpath('//bridge').attribute('name').content
+    end
+    if doc.at_xpath('//forward')
+      forward = doc.at_xpath('//forward')
+      if forward.attribute('mode')
+        definition[:forward_mode] = forward.attribute('mode').content
+      end
+      if forward.attribute('dev')
+        definition[:forward_dev] = forward.attribute('dev').content
+      end
+      if forward.xpath('//interface')
+        definition[:forward_interfaces] = []
+        for int in forward.xpath('//interface')
+          definition[:forward_interfaces].push(int.attribute('dev').content)
+        end
+      end
+    end
+    if doc.xpath('//ip')
+      definition[:ip] = []
+      definition[:ipv6] = []
+      for ip in doc.xpath('//ip')
+        data = {}
+        for setting in ['address', 'netmask', 'prefix']
+          if ip.attribute(setting)
+            data[setting] = ip.attribute(setting).content
+          end
+        end
+        if ip.at_xpath('//dhcp')
+          data[:dhcp] = {}
+          for setting in ['start', 'end']
+            data[:dhcp][setting] = ip.at_xpath('//dhcp/range').attribute(setting).content
+          end
+          if ip.at_xpath('//dhcp/bootp')
+            data[:dhcp][:bootp_file] = ip.at_xpath('//dhcp/bootp').attribute('file').content
+          end
+        end
+        if ip.attribute('family') and ip.attribute('family').content == 'ipv6'
+          definition[:ipv6].push(data)
+        else
+          definition[:ip].push(data)
+        end
+      end
+    end
+    if doc.at_xpath('//mac')
+      definition[:mac] = doc.at_xpath('//mac').attribute('address').content
+    end
+    return definition
+  end
+
+  def self.instances 
+    networks = []
+    for net in $conn.list_all_networks() do
+      hash = parse_network(net)
+      hash[:autostart] = net.autostart?
+      networks << new(hash)
+    end
+    return networks
+  end
+
+  def self.prefetch(resources)
+    instances.each do |prov|
+      if resource = resources[prov.name]
+        resource.provider = prov
+      end
+    end
+  end
+
+  def create
+ #   @property_hash = @resource
```

Comment:
The commented out line is leftover from some experimentation, I'll remove it. The ```create``` function must exist, even if it's empty, otherwise puppet will throw an error.